README.md
## Btcli4j

![GitHub Workflow Status](https://img.shields.io/github/workflow/status/clightning4j/btcli4j/Java%20CI?style=flat-square)
![Esplora support](https://img.shields.io/badge/esplora-supported-gren?style=flat-square)
![Pruning mode](https://img.shields.io/badge/prune-supported-gren?style=flat-square)

It is a [c-lightning](https://lightning.readthedocs.io/index.html) plugin to override Bitcoin backend plugin with [esplora](https://github.com/Blockstream/esplora) 
powered by [Blockstream](https://blockstream.com/).

But this plugin is designed make the process to run c-lightning with bitcoind in pruning mode, the idea of this plugin was
described inside this [issue by lightningd](https://github.com/lightningd/plugins/issues/112).

It is designed with mediator patter and each command inside the mediator can use esplora or bitcoind or a both in some cases.

So, the plugin will support the complete backend with esplora (only) and the complete backend only with bitcoind or a join of the two option!

## Network supported

- [X] Bitcoin (Mainet (Be cautious) and Testnet)
- [X] Liquid
- [X] Bitcoin Core pruning mode (Mainet (Be cautious) and Testnet)
- [X] Bitcoin Core remote call (have bitcoin in another location of your lightningd node) (Mainet (Be cautious) and Testnet)
- [ ] Litecoin
- [ ] Liquid

## Status

The plugin is in Beta stage, this mean that is waiting for the User Review and testing, good tested on bitcoin testnet
but not on mainet, because I have put inside only small amount of found that available at the moment, if you want support 
the project consider to look in the section [Support]()

If you want test it all feedback are welcome! Feel free to open an issue or a PR or in addition, you can send
email to [vincenzopalazzodev@gmail.com](mailito://vincenzopalazzodev@gmail.com)

## Install
Java produces a jar and c-lightning needs a bash script to run it! 
The gradle script compiles the plugin and generate a bash script with the command `./gradlew createRunnableScript`

After the gradle process, you have the jar inside the `build/libs/btcli4j-all.jar` and the script `btcli4j-gen.sh` 
in the root directory of the project.

Now you can put the gradle script inside the c-lightning plugin directory

### How bash script look like

The contains only the command to run the jar with java, in my cases the script contains the following result>

```bash
#!/bin/bash
/usr/lib/jvm/jdk-13.0.2/bin/java -jar /home/vincent/Github/btcli4j/build/libs/btcli4j-all.jar
```

In this case, you can move this bash script inside your `./lightning/plugins` the directory or you can add the plugin to the file conf
with `plugin=/PATH/bash/file` or use the command line `--plugin=/path/bash/file`.

## Plugin parameter

- btcli4j-endpoint: A string option that is the end point url,by default it is the [esplora](https://github.com/Blockstream/esplora/blob/master/API.md) end point, 
- it is the end point it is different the proxy with tor is lost from the automatic configuration of the plugin
- btcli4j-waiting: A integer that is the base waiting time where the plugin will apply the increase waiting time in case of error with the end point.
- btcli4j-proxy: Indicate the proxy address, by default is: "127.0.0.1:9050".
- btcli4j-proxy-disabled: A flag option, it helps the user to disabled the cln proxy setting if enabled.
- btcli4j-pruned: A flag option that tell to the plugin to ran in pruning mode
- bitcoin-rpcurl: A string option that is the url of Bitcoin RPC interface, usually `http://127.0.0.1:8332` for mainet and `http://127.0.0.1:18332` for testnet
- bitcoin-rpcuser: A string option that is the user to make the login in Bitcoin core
- bitcoin-rpcpassword: A string option that is the password to make the login in Bitcoin core

P.S: some of Bitcoin propriety are the same of the C-lightning propriety, these mean that yuo need to redefine only a ``bitcoin-rpcurl`` propriety, and put
at the startup of the lightning node the flag `btcli4j-pruned`

### Esplora API config

A complete example to run with Esplora API is reported below

```bash
lightningd --disable-plugin bcli
```

The command above, run the lightningd with the lightnind configuration, this mean that you don't need to add some 
additional configuration to the command line.

However, you can use the command line to customize the plugin behaviors.

### Bitcoin Pruning mode config

```bash
lightningd --disable-plugin bcli --btcli4j-pruned
```

With the following clightning config

```bash
...
bitcoin-rpcuser=user
bitcoin-rpcpassword=pass
bitcoin-rpcurl=http://127.0.0.1:18332
....
```

## Code Style
[![ktlint](https://img.shields.io/badge/code%20style-%E2%9D%A4-FF4081.svg)](https://ktlint.github.io/)

> We live in a world where robots can drive a car, so we shouldn't just write code, we should write elegant code.

This repository use [ktlint](https://github.com/pinterest/ktlint) to maintains the code of the repository elegant, so 
before submit the code check the Kotlin format with the following command on the root of the directory

```bash
./gradlew formatKotlin
```

## Built with

- [JRPClightning](https://github.com/vincenzopalazzo/JRPClightning)
- [lite-bitcoin-rpc](https://github.com/clightning4j/lite-bitcoin-rpc)

## Support
If you like the library and want to support it, please considerer to donate with the following system


- [liberapay.com/vincenzopalazzo](https://liberapay.com/vincenzopalazzo)
- [3BQ8qbn8hLdmBKEjt1Hj1Z6SiDsnjJurfU](bitcoin:3BQ8qbn8hLdmBKEjt1Hj1Z6SiDsnjJurfU)
- [Github support](https://github.com/sponsors/vincenzopalazzo)

## License

<div align="center">
  <img src="https://opensource.org/files/osi_keyhole_300X300_90ppi_0.png" width="150" height="150"/>
</div>

 It is a c-lightning plugin to override Bitcoin backend plugin with esplora.

 Copyright (C) 2020-2021 Vincenzo Palazzo vincenzopalazzodev@gmail.com
 
 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License along
 with this program; if not, write to the Free Software Foundation, Inc.,
 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


CHANGELOG.md
# v0.0.7

## Fixed
- fix the estimation of the fee in pruning mode ([commit](https://github.com/clightning4j/btcli4j/commit/76d9823120cba399591766120a569bbb63551a15)). @vincenzopalazzo 29-03-2023
- coffe script ([commit](https://github.com/clightning4j/btcli4j/commit/de360fa8e24e31e693c8636a32eac7ce07e9cea8)). @vincenzopalazzo 08-03-2023
- coffe script ([commit](https://github.com/clightning4j/btcli4j/commit/c233640b005600dd4f9b41c9b49a81b8e1733a9c)). @vincenzopalazzo 08-03-2023

## Added
- support coffee manifest ([commit](https://github.com/clightning4j/btcli4j/commit/e006d838477b149d650c36eb1e065cc40813bd77)). @vincenzopalazzo 08-03-2023


# v0.0.6

## Fixed
- fix the possibility to disable the cln proxy ([commit](https://github.com/clightning4j/btcli4j/commit/99046cd04609a8a0c9dfcbb56e1a0de57150a9aa)). @vincenzopalazzo 28-01-2023


# v0.0.5

## Fixed
- upgrade the java library with the upstream ([commit](https://github.com/clightning4j/btcli4j/commit/d04d0a03a54d24fcaedbd1021eba2ce566dffe36)). @vincenzopalazzo 14-01-2023


# v0.0.4

## Fixed
- send raw transaction cln response ([commit](https://github.com/clightning4j/btcli4j/commit/7e39245cf4c1b0c623a361550dffb2d3a76e0e69)). @vincenzopalazzo 16-11-2022
- use 6 block of estimation fee ([commit](https://github.com/clightning4j/btcli4j/commit/d2bca9236b17db03003c1d6fbc50798d47db5825)). @vincenzopalazzo 16-11-2022
- if a error is throws from bitcoind return the error to cln ([commit](https://github.com/clightning4j/btcli4j/commit/7ab269a2407e0d721d0fce2d66db29d9cb122454)). @vincenzopalazzo 16-11-2022
- Update gitignore. ([commit](https://github.com/clightning4j/btcli4j/commit/d281e041159ce500dbec237c3b5d26c5d72526d9)). @swaptr 28-05-2022


# v0.0.3

## Fixed
- if a error is throws from bitcoind return the error to cln ([commit](https://github.com/clightning4j/btcli4j/commit/7ab269a2407e0d721d0fce2d66db29d9cb122454)). @vincenzopalazzo 16-11-2022
- Update gitignore. ([commit](https://github.com/clightning4j/btcli4j/commit/d281e041159ce500dbec237c3b5d26c5d72526d9)). @swaptr 28-05-2022


src/main/kotlin/io/vincenzopalazzo/btcli4j/Plugin.kt
/**
 *  C-lightning plugin to override Bitcoin backend plugin.
 *  Copyright (C) 2020-2021 Vincenzo Palazzo vincenzopalazzodev@gmail.com
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
package io.vincenzopalazzo.btcli4j

import io.vincenzopalazzo.btcli4j.control.MediationMethod
import io.vincenzopalazzo.btcli4j.util.HttpRequestFactory
import io.vincenzopalazzo.btcli4j.util.OptionsManager
import io.vincenzopalazzo.btcli4j.util.PluginManager
import jrpc.clightning.annotation.PluginOption
import jrpc.clightning.annotation.RPCMethod
import jrpc.clightning.plugins.CLightningPlugin
import jrpc.clightning.plugins.log.PluginLog
import jrpc.service.converters.jsonwrapper.CLightningJsonObject

/**
 * @author https://github.com/vincenzopalazzo
 */
class Plugin : CLightningPlugin() {

    @PluginOption(
        name = "btcli4j-endpoint",
        description = "This option give information on rest end point url, by default it is the esplora url. " +
            "If you specify this URL you are losing the autoconfiguration of the tor proxy in case of proxy enabled with lightningd.",
        defValue = "https://blockstream.info",
        typeValue = "string"
    )
    private var endpointUrl: String = "https://blockstream.info"

    @PluginOption(
        name = "btcli4j-waiting",
        description = "This option give information on the waiting time in case of error with the endpoint connection. " +
            "In case of failure this value is the base value where the plugin start to use the increase waiting time rules.",
        defValue = "60000",
        typeValue = "int"
    )
    private var waitingTime: Int = 60000

    @PluginOption(
        name = "btcli4j-proxy",
        description = "This option give information on proxy enabled, by default set on tor proxy",
        defValue = "127.0.0.1:9050",
        typeValue = "string"
    )
    private var proxy: String = "127.0.0.1:9050"

    @PluginOption(
        name = "btcli4j-proxy-disabled",
        description = "This option give disable the proxy inside plugin. By default is false and use the cln settings",
        defValue = "false",
        typeValue = "flag"
    )
    private var proxyDisabled: Boolean = false

    @PluginOption(
        name = "bitcoin-rpcpassword",
        description = "Bitcoin RPC password",
        defValue = "",
        typeValue = "string"
    )
    private var bitcoinRpcPass: String = ""

    @PluginOption(
        name = "bitcoin-rpcuser",
        description = "Bitcoin RPC user",
        defValue = "",
        typeValue = "string"
    )
    private var bitcoinRpcUser: String = ""

    @PluginOption(
        name = "bitcoin-rpcurl",
        description = "Base URL Bitcoin RPC",
        defValue = "",
        typeValue = "string"
    )
    private var bitcoinBaseUrl: String = ""

    @PluginOption(
        name = "btcli4j-pruned",
        description = "Tell to BTC4J plugin to use the pruned mode",
        defValue = "false",
        typeValue = "flag"
    )
    private var prunedMode = false

    private var pluginInit = false

    @RPCMethod(name = "getchaininfo", description = "getchaininfo to fetch the data from blockstream.info")
    fun getChainInfo(plugin: CLightningPlugin, request: CLightningJsonObject, response: CLightningJsonObject) {
        this.configurePluginInit()
        MediationMethod.runCommand("getchaininfo", plugin, CLightningJsonObject(request["params"].asJsonObject), response)
    }

    @RPCMethod(name = "estimatefees", description = "estimatefees to fetch the feed estimation from blockstream.info")
    fun estimateFees(plugin: CLightningPlugin, request: CLightningJsonObject, response: CLightningJsonObject) {
        MediationMethod.runCommand("estimatefees", plugin, CLightningJsonObject(request["params"].asJsonObject), response)
    }

    @RPCMethod(name = "getrawblockbyheight", description = "getrawblockbyheight to fetch actual blockchain height from blockstream.info")
    fun getRawBlockByHeight(plugin: CLightningPlugin, request: CLightningJsonObject, response: CLightningJsonObject) {
        MediationMethod.runCommand("getrawblockbyheight", plugin, CLightningJsonObject(request["params"].asJsonObject), response)
    }

    @RPCMethod(name = "getutxout", description = "getutxout to fetch a utx with {txid} and {vout} from blockstream.info")
    fun getUtxOut(plugin: CLightningPlugin, request: CLightningJsonObject, response: CLightningJsonObject) {
        MediationMethod.runCommand("getutxout", plugin, CLightningJsonObject(request["params"].asJsonObject), response)
    }

    @RPCMethod(name = "sendrawtransaction", description = "sendrawtransaction to publish a new transaction with blockstream.info")
    fun sendRawTransaction(plugin: CLightningPlugin, request: CLightningJsonObject, response: CLightningJsonObject) {
        MediationMethod.runCommand("sendrawtransaction", plugin, CLightningJsonObject(request["params"].asJsonObject), response)
    }

    private fun configurePluginInit() {
        // TODO: The plugin have some proxy enabled and proxy url propriety to give the possibility to the user
        // to specify a custom proxy only of the http client used by the plugin. Implement this view.
        val optionsManager = OptionsManager(endpointUrl, waitingTime, null, !this.proxyDisabled, proxy)
        if (!pluginInit && configs.isProxyEnabled && !this.proxyDisabled) {
            val proxyIp = configs.proxy.address
            val proxyPort = configs.proxy.port
            val torVersion = when (configs.proxy.type) {
                "torv3" -> 3
                "torv2" -> 2
                else -> if (configs.isTorv3) 3 else null
            }
            optionsManager.proxyEnabled = configs.isProxyEnabled
            optionsManager.proxyUrl = "%s:%d".format(proxyIp, proxyPort)
            optionsManager.torVersion = torVersion
            log(PluginLog.INFO, "btcli4j: proxy enable at ${optionsManager.proxyUrl}")
            log(PluginLog.INFO, "btcli4j: proxy type is ${configs.proxy.type}")
            log(PluginLog.INFO, "btcli4j: tor version is $torVersion")
        }

        if (prunedMode && bitcoinRpcPass.trim().isNotEmpty() && bitcoinRpcUser.trim().isNotEmpty()) {
            PluginManager.instance.bitcoinPass = bitcoinRpcPass
            PluginManager.instance.bitcoinUser = bitcoinRpcUser
            PluginManager.instance.baseBitcoinUrl = bitcoinBaseUrl
            PluginManager.instance.prunedMode = true
        }
        HttpRequestFactory.initHttpClient(optionsManager)
        pluginInit = true
    }
}
